src/Collection/Resources/assets/src/controller.ts

```diff
@@ -0,0 +1,110 @@
+    }
+
+    createButton(collectionEl: HTMLElement, buttonType: ButtonType): HTMLElement {
+        const attributeName = `${ButtonType[buttonType].toLowerCase()}ButtonTemplateId`;
+        const buttonTemplateID = collectionEl.dataset[attributeName] ?? (this as any)[`${attributeName}Value`];
+        if (buttonTemplateID && 'content' in document.createElement('template')) {
+            // Get from template
+            const buttonTemplate = document.getElementById(buttonTemplateID) as HTMLTemplateElement | null;
+            if (!buttonTemplate) throw new Error(`template with ID "${buttonTemplateID}" not found`);
+
+            const fragment = buttonTemplate.content.cloneNode(true) as DocumentFragment;
+            if (1 !== fragment.children.length)
+                throw new Error('template with ID "${buttonTemplateID}" must have exactly one child');
+
+            return fragment.firstElementChild as HTMLElement;
+        }
+
+        // If no template is provided, create a raw HTML button
+        const button = document.createElement('button') as HTMLButtonElement;
+        button.type = 'button';
+        button.textContent = buttonType === ButtonType.Add ? 'Add' : 'Delete';
+
+        return button;
+    }
+
+    addItem(collectionEl: HTMLElement) {
+        const currentIndex = (collectionEl.dataset as CollectionDataset).currentIndex;
+        (collectionEl.dataset.currentIndex as unknown as number)++;
+
+        const collectionNamePattern = collectionEl.id.replace(/_/g, '(?:_|\\[|]\\[)');
+
+        const prototype = (collectionEl.dataset.prototype as string) // We're sure that dataset.prototype exists, because of the CSS selector used in connect()
+            .replace('__name__label__', currentIndex)
+            .replace(new RegExp(`(${collectionNamePattern}(?:_|]\\[))__name__`, 'g'), `$1${currentIndex}`);
+
+        const fakeEl = document.createElement('div');
+        fakeEl.innerHTML = prototype;
+        const itemEl = fakeEl.firstElementChild as HTMLElement;
+
+        this.connectCollection(itemEl);
+
+        this.addDeleteButton(collectionEl, itemEl);
+
+        const items = this.getItems(collectionEl);
+        items.length ? items[items.length - 1].insertAdjacentElement('afterend', itemEl) : collectionEl.prepend(itemEl);
+    }
+
+    addAddButton(collectionEl: HTMLElement) {
+        const addButton = this.createButton(collectionEl, ButtonType.Add);
+        addButton.onclick = (e) => {
+            e.preventDefault();
+            this.addItem(collectionEl);
+        };
+        collectionEl.appendChild(addButton);
+    }
+
+    addDeleteButton(collectionEl: HTMLElement, itemEl: HTMLElement) {
+        const deleteButton = this.createButton(collectionEl, ButtonType.Delete);
+        deleteButton.onclick = (e) => {
+            e.preventDefault();
+            itemEl.remove();
+        };
+        itemEl.appendChild(deleteButton);
+    }
```

Comment:
In #90 where was discussed that it would be nice that the buttons are rendered by the symfony form theme. That make it special for design systems easier so the can directly use there already designed button_widget and bootstrap and other themes already ships with a already good looking button. currently buttons are here if not explicit a template is defined rendered via this js component. in #400 I used the twig extension to add the buttons already to the twig and to the prototype, which did make also the JS a lot smaller which I think is great from maintainance point of view.

Reply:
My long-term goal is to totally remove `data-prototype` from core, generate `<template>` element instead (which are better suited for this use case), and provide default buttons directly in the core form themes. However, during the deprecation process, we'll need to support this and `data-prototype`) for the "plug & play" feature until Symfony 7.

I'll open the PR on core in the next few days.

Reply:
Agree still that `template` are better suited for the `data-prototype. Still I recommed to have the buttons rendered already in `twig` because adding them via `JS` will trigger else a `Layout shift` and google and lighthouse hate layout shifts.

Reply:
Sure, that's the plan (and will be the default in Symfony 7!)